R/rvar-summaries-within-draws.R

```diff
@@ -0,0 +1,160 @@
+#'
+#' Each function defined here corresponds to the base function of the same name
+#' without the `rvar_` prefix (e.g., `rvar_mean()` calls `mean()` under the hood, etc).
+#'
+#' @return
+#' An [`rvar`] of length 1 (or in the case of `range()`, length 2) with the same number
+#' of draws as the input rvar(s) containing the summary statistic computed within
+#' each draw of the input rvar(s).
+#'
+#' @examples
+#'
+#' set.seed(5678)
+#' x = rvar_rng(rnorm, 4, mean = 1:4, sd = 2)
+#'
+#' # These will give similar results to mean(1:4),
+#' # median(1:4), sum(1:4), prod(1:4), etc
+#' rvar_mean(x)
+#' rvar_median(x)
+#' rvar_sum(x)
+#' rvar_prod(x)
+#'
+#' @seealso [rvar-summaries-over-draws] for summary functions across draws (e.g. expectations).
+#' [rvar-functions] for density, CDF, and quantile functions of random variables.
+#' @family rvar-summaries
+#' @name rvar-summaries-within-draws
+#' @export
+rvar_mean <- function(..., na.rm = FALSE) summarise_rvar_within_draws(c(...), mean, na.rm = na.rm)
+
+
+# numeric summaries -------------------------------------------------------
+
+#' @rdname rvar-summaries-within-draws
+#' @export
+rvar_median <- function(..., na.rm = FALSE) summarise_rvar_within_draws(c(...), median, na.rm = na.rm)
+
+#' @rdname rvar-summaries-within-draws
+#' @export
+rvar_sum <- function(..., na.rm = FALSE) summarise_rvar_within_draws(c(...), sum, na.rm = na.rm)
+
+#' @rdname rvar-summaries-within-draws
+#' @export
+rvar_prod <- function(..., na.rm = FALSE) summarise_rvar_within_draws(c(...), prod, na.rm = na.rm)
+
+#' @rdname rvar-summaries-within-draws
+#' @export
+rvar_min <- function(..., na.rm = FALSE) summarise_rvar_within_draws(c(...), min, na.rm = na.rm)
+
+#' @rdname rvar-summaries-within-draws
+#' @export
+rvar_max <- function(..., na.rm = FALSE) summarise_rvar_within_draws(c(...), max, na.rm = na.rm)
+
+
+# spread ------------------------------------------------------------------
+
+#' @rdname rvar-summaries-within-draws
+#' @export
+rvar_sd <- function(..., na.rm = FALSE) summarise_rvar_within_draws(c(...), sd, na.rm = na.rm)
+
+#' @rdname rvar-summaries-within-draws
+#' @export
+rvar_var <- function(..., na.rm = FALSE) {
+  # var() is silly and gives the covariance matrix instead of the variance
+  # when dim(x) == 2, so convert to a vector to avoid this
+  summarise_rvar_within_draws(c(...), function(x) var(as.vector(x), na.rm = na.rm))
+}
+
+#' @rdname rvar-summaries-within-draws
+#' @export
+rvar_mad <- function(..., constant = 1.4826, na.rm = FALSE, low = FALSE, high = FALSE) {
+  summarise_rvar_within_draws(c(...), mad, constant = constant, na.rm = na.rm, low = low, high = high)
+}
+
+
+# range -------------------------------------------------------------------
+
+#' @rdname rvar-summaries-within-draws
+#' @export
+rvar_range <- function(..., na.rm = FALSE) summarise_rvar_within_draws(c(...), range, na.rm = na.rm, .transpose = TRUE)
```

Comment:
since we know how to handle range already with more than 1 ouput, could it make sense to also add `rvar_quantile`?

Reply:
Good idea. I suppose we could do a quantile2 implementation as well? I've added them to the list in #117 (https://github.com/stan-dev/posterior/issues/117#issuecomment-811484159) and will handle them in the next round of changes (not this PR).